### ISSUE TITLE
macOS 10.15 Compatibility

### ISSUE THREAD

**Issue by DomenicBianchi01** [leezer3/OpenBVE#384]
Are there any plans to update OpenBVE to be able to run on macOS 10.15 (Catalina)? Currently the program crashes on start up. Seems like it's because it is using the 32-bit version of Mono but macOS 10.15 no longer supports 32-bit frameworks/programs.

**Comment by s520** [comment]
Have you tried the options below?
`--arch=32`
https://www.mono-project.com/docs/about-mono/supported-platforms/macos/

**Comment by leezer3** [comment]
Interesting question.

On Windows (for legacy reasons) we're stuck on 32-bit.

I was under the impression that Catalina only removed some default 32-bit libraries, as opposed to totally killing 32-bit support dead.
The build cpuld be switched to 64-bit if Mono gets a fully working 64-bit winforms stack.

I'd be intetested to see the error text to start with.
I don't have a real Mac, let alone one with Catalina so this may be an issue, bit it can certainly be looked into.

**Comment by DomenicBianchi01** [comment]
@s520 macOS 10.15 no longer supports 32-bit architectures at all.

@leezer3 When I try to open OpenBVE I get the following error in the console: `Unsupported 32-bit executable: "/Library/Frameworks/Mono.framework/Versions/Current/bin/mono32"`

**Comment by leezer3** [comment]
Ouch, that's not good at all; They've definitely dumped 32-bit entirely (terminally stupid, but there we go...)

I'm surprised mono didn't provide a 64-bit shim for mono32 though :(

Digging around in the Mono issues also gives me a link to this:
https://github.com/tilmanginzel/alfred-bluetooth-workflow/issues/9
That suggests that unless Gatekeeper is disabled (not simply bypassed), we'll also dump out, as we're not code-signed.

I will try the x64 version of Mono on a Linux box some time next week and see if we can at least launch with that, but the main menu makes heavy use of legacy winforms, which means that might be problematic at best.

TLDR:

- I suspect Catalina compatibility will not happen any time soon, sorry :(
- None of our code requires 32-bit, but on Windows we are tied to this for legacy reasons, which means it's rather hard to change.
- Any 64-bit support under Mono would require a working Winforms implementation. Last time I looked, Mono didn't have this. .Net Core 3.0 *might* get this, but that's not current yet.

**Comment by DomenicBianchi01** [comment]
@leezer3 Regarding the code-signing issue you linked, you can actually bypass that. You just have to go to System Preferences -> Security and in there is a button to tell macOS it's ok to open this program.

I think Apple said someday everything will need to be code-signed but that day has not come yet.

**Comment by s520** [comment]
I tested with macOS 10.15 Beta 8.
When I try to start mono32, the error "zsh: bad CPU type in executable: mono32" occurs.
.NET Core 3.0 winforms seems to be Windows only, so we may need to use a framework other than winforms ...

**Comment by leezer3** [comment]
Related- This seems to be the official Mono standpoint at present:
https://github.com/mono/mono/issues/16890

I'm afraid nothing is likely to change significantly in the near future at our end without someone stepping up to rebuild the main menu in an alternate manner :(

For version 1.8, the target will likely to be to move the main sim code to x64 via the PR I've linked above or a variant upon it, but the main game start menu is a very complicated bit of Windows forms code.
We *might* be able to produce a simplified openGL main menu to allow loading of routes and trains etc. but it wouldn't have anywhere near the polish of what we have at present.

**Comment by s520** [comment]
Creating a menu with OpenGL will not be easy.
I suggest Xamarin.Forms as the migration destination.
This is easier to make menus.

**Comment by DomenicBianchi01** [comment]
@leezer3 @s520 It's probably worth noting that OpenGL is deprecated as of macOS 10.14 (Mojave). For now Apple hasn't announced any plans to remove OpenGL from the macOS, however, since it is deprecated it is very likely Apple will be removing it (in favour of Metal) within the next few years.

As a result, I think it would be wise to not use OpenGL (just to avoid creating more work at a later date).

PS. I appreciate the time you've spent looking into the macOS 10.15 compatibility issue. Thank-you!

**Comment by leezer3** [comment]
Now that is insane :/
If openGL goes, at present we're stuffed. 
Some progress has been made, and I think we've now got much of the foundations in place for alternate render paths, but at the minute we're heavily reliant on openGL intermediate mode, and I can't see that changing soon......

TBQH it's no big deal investigating something like this, as if nothing else it provided the kick to start thinking about getting x64 in place properly.

### Xamarin.Forms vs. openGL-
Been doing a little playing with this. The Xamarin forms seem to be a little easier to implement in the short term, but require a Mac to actually run the preview successfuly and test, which I don't have constant access to. It also means that a Mac would require a 100% Mac specific build, as it's somehow building an embedded Coaco app that does the window drawing.

openGL has the major advantage that it'll work on anything, but requires considerable work to be anything close to usable.
A simple route selector can be hacked together using the current menu code, but this isn't brilliant. Will try and stuff up a branch with this on.

**Comment by s520** [comment]
No.
If you use Xamarin.Forms.Platform.GTK, you don't need a macOS specific build.
https://docs.microsoft.com/en-gb/xamarin/xamarin-forms/platform/other/gtk?tabs=windows

**Comment by leezer3** [comment]
Interesting, that could well work.

Will still play with the openGL menu idea though.
With the commit above (It fixes some generic enough issues to go into the mainline builds), and some minor hacking in the gamewindow init code, I've got the in-game menu working as soon as the program loads. 
Sure, that isn't too helpful as it stands, but we can define and pop menus at our leisure. Will keep playing :)

**Comment by leezer3** [comment]
The openGL idea is starting to take shape.

This branch is where I'm working at the minute:
https://github.com/leezer3/OpenBVE/compare/OpenGLMenu

### Functional:
* Lists CSV / RW route files.
* Allows browsing up and down the folder structure on the same drive.
### TODO:
* Hook up so that we can actually launch a route.
* Train selector- Likely to be similar to the route selector. e.g. a list of available folders / files.
* Left-align the menu to half-screen width to allow for the fancy business.
### Fancy business (later!):
* Route / train images (not too difficult)
* Route description if possible. (Text wrapping in openGL may be a PITA)

**Comment by mattwaltbriggs** [comment]
Is M1 support being considered for openBVE at this time?

**Comment by leezer3** [comment]
Not at present, sorry.
I don't have regular access to a 'real' 64-bit Mac, let alone an M1.

---
Current situation FWIW:
The next step in getting closer to this will be merging this branch to give us native 64-bit builds:
https://github.com/leezer3/OpenBVE/tree/WinPluginProxy
This is on the list of things to be done soon-ish.....
With that, IIRC the main game and the viewers work.
Unfortunately however, the main launcher menu is a somewhat complex WinForm. These aren't supported on 64-bit Apple by Mono:
https://github.com/mono/mono/issues/6701

Route Viewer / Object Viewer *might* work on an M1 if you merged the above branch and built using 64-bit, but I'm afraid you'd be on your own there.

**Comment by s520** [comment]
The fundamental solution is to create a native 64-bit version, but the current 32-bit version may work on [CrossOver](https://github.com/Gcenx/homebrew-wine).
Too bad I don't have an actual machine to verify that...

Also, there was a report in the Japanese community that OpenBVE worked on a Nintendo Switch (Tegra X1) with Linux installed.
(However, according to the person who reported it, it was very slow and could not be used regularly.)
This suggests that if it works on Intel Mac, it will also work on M1 Mac.

**Comment by leezer3** [comment]
Now that's interesting.
IIRC there's also been someone running it on a Pi now that you mention it.

The new renderer is a lot more likely to work / be a sensible speed, but it really depends on what the OpenGL drivers support.
I'm guessing the Switch was probably using MESA emulation, which would definitely be slow.

**Comment by OmegaZeroKiller** [comment]
> Not at present, sorry.
> I don't have regular access to a 'real' 64-bit Mac, let alone an M1.
> 
> Route Viewer / Object Viewer _might_ work on an M1 if you merged the above branch and built using 64-bit, but I'm afraid you'd be on your own there.

Greetings:
OmegaZeroKiller from the Hong Kong community here, and I have recently purchased a M1 Mac Mini for OpenBVE performance testing and evaluation; I also have access to 64-bit Intel Macs for further testing.

However, I am relatively new to Unix-like OSes, and could use some guidance in finding a way to launch OpenBVE successfully on Mac OS X / macOS without using the standard installer from the main OpenBVE Project site; other than that setback, I am more than happy to conduct testing for you.

**Comment by leezer3** [comment]
@mattwaltbriggs @DomenicBianchi01 
The OpenGL menu code has *finally* been merged to the daily builds.
Builds from the 20th of June onwards should now be usable.
https://vps.bvecornwall.co.uk/OpenBVE/Builds/OpenBVE-2021-07-20.dmg

Whilst this is a relatively basic solution (there's no real access to options etc.), it's workable, and will hopefully be improved in the longer term.
From comments in the linked PR, M1 works just fine with this. 

I'm going to close this issue as the basic incompatability problem has been solved.
This will be in the public 1.9.0.0, but I don't want to give any timescales at the minute.

**Comment by plutommm** [comment]
It seems like this version [https://vps.bvecornwall.co.uk/OpenBVE/Builds/OpenBVE-2021-07-20.dmg](url) of OpenBVE is not working on my new M1 Mac... I would like to know the requirements needed to run this version.
